plumbing/format/diff/unified_encoder.go

```diff
@@ -0,0 +1,345 @@
+package diff
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"gopkg.in/src-d/go-git.v4/plumbing"
+)
+
+const (
+	diffInit = "diff --git a/%s b/%s\n"
+
+	chunkInit  = "@@ -%s +%s @@%s\n"
+	chunkCount = "%d,%d"
+
+	noFilePath = "/dev/null"
+	aDir       = "a/"
+	bDir       = "b/"
+
+	fPath  = "--- %s\n"
+	tPath  = "+++ %s\n"
+	binary = "Binary files %s and %s differ\n"
+
+	addLine    = "+%s\n"
+	deleteLine = "-%s\n"
+	equalLine  = " %s\n"
+
+	oldMode         = "old mode %o\n"
+	newMode         = "new mode %o\n"
+	deletedFileMode = "deleted file mode %o\n"
+	newFileMode     = "new file mode %o\n"
+
+	renameFrom     = "from"
+	renameTo       = "to"
+	renameFileMode = "rename %s %s\n"
+
+	indexAndMode = "index %s..%s %o\n"
+	indexNoMode  = "index %s..%s\n"
+)
+
+var ErrBothFilesEmpty = errors.New("both files are empty")
+
+// UnifiedEncoder encodes an unified diff into the provided Writer. There are some unsupported features:
+// - Similarity index is not supported
+// - Sort hash representation is not supported
+type UnifiedEncoder struct {
+	io.Writer
+	CtxLines int
+}
+
+func (e *UnifiedEncoder) Encode(patch Patch) error {
+	if err := e.printMessage(patch.Message()); err != nil {
+		return err
+	}
+
+	for _, p := range patch.FilePatches() {
+		f, t := p.Files()
+		if err := e.header(f, t, p.IsBinary()); err != nil {
+			return err
+		}
+
+		g := newHunksGenerator(p.Chunks(), e.CtxLines)
+		for _, c := range g.Generate() {
+			if err := e.prints(c.String()); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (e *UnifiedEncoder) printMessage(message string) error {
+	isEmpty := message == ""
+	hasSuffix := strings.HasSuffix(message, "\n")
+	msg := message
+	if !isEmpty && !hasSuffix {
+		msg = msg + "\n"
+	}
+
+	return e.prints(msg)
+}
+
+func (e *UnifiedEncoder) header(from, to File, isBinary bool) error {
+	var fromPath, toPath string
+	var initLine, modesLine, indexLine string
+	var equals bool
+
+	switch {
+	case from == nil && to == nil:
+		return ErrBothFilesEmpty
+	case from != nil && to != nil:
+		initLine = fmt.Sprintf(diffInit, from.Path(), to.Path())
```

Comment:
I don't fully understand what are you trying to say. You mean that, instead of do this:
```
modesLines = fmt.Sprintf(oldMode+newMode, from.Mode(), to.Mode())
```
should do this?:
```
buf.WriteString("old mode ")
buf.WriteString(from.Mode())
buf.WriteString("\n")
buf.WriteString("new mode ")
buf.WriteString(to.Mode())
buf.WriteString("\n")
```